llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,234 @@
+// picked up by loop rotate to remove unncessary loop guards.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-iter-count-assumptions"
+
+using namespace llvm;
+namespace {
+
+struct BundledSCEV {
+  /// AddRec Scalar Evolution Expression of the IV
+  const SCEVAddRecExpr *AddRecExpr = nullptr;
+  /// Operand Index of the IV in the compare Instruction
+  int IVOperandIndex;
+
+  BundledSCEV(const SCEVAddRecExpr *AddRecExpr, int IVOperandIndex)
+      : AddRecExpr(AddRecExpr), IVOperandIndex(IVOperandIndex) {}
+};
+
+std::string getFunctionAndBlockNames(BasicBlock &BB) {
+  return BB.getParent()->getName().str() + " " + BB.getName().str();
+}
+
+/// get AddRec Scalar Evolution Expression, if it can be found in Op
+const SCEVAddRecExpr *getAddRecSCEV(Value *Op, ScalarEvolution &SE) {
+  if (!SE.isSCEVable(Op->getType()))
+    return nullptr;
+
+  return dyn_cast<SCEVAddRecExpr>(SE.getSCEV(Op));
+}
+
+/// get the BundledSCEV of the loop IV which consists of the AddRec Scalar
+/// Evolution Expression and the operand index of the IV in LoopCmpInstr
+std::optional<BundledSCEV> getLoopCmpIVAddRecExpr(const ICmpInst &LoopCmpInstr,
+                                                  ScalarEvolution &SE) {
+
+  const SCEVAddRecExpr *S0 = getAddRecSCEV(LoopCmpInstr.getOperand(0), SE);
+  const SCEVAddRecExpr *S1 = getAddRecSCEV(LoopCmpInstr.getOperand(1), SE);
+
+  if (S0 && S1) {
+    LLVM_DEBUG(
+        dbgs() << "LoopIterCountAssumptions-Warning: Could not extract "
+                  "SCEVAddRecExpr.\nEncountered two ADDRec "
+                  "expressions in the Loop Compare instruction.\nCurrently it "
+                  " is expected that the compare consists of the IV and a "
+                  "static bound, that is loop invariant.");
+    return std::nullopt;
+  }
+
+  if (S0)
+    return std::optional(BundledSCEV(S0, 0));
+
+  if (S1)
+    return std::optional(BundledSCEV(S1, 1));
+
+  LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: Could not extract "
+                       "SCEVAddRecExpr.\nEncountered no ADDRec "
+                       "expressions in the Loop Compare Instruction.");
+  return std::nullopt;
+}
+
+/// return the Branch Compare Instruction of CurrentLoop if the loop is well
+/// formed and this pass can process the predicate
+ICmpInst *getLoopCmpInst(Loop &CurrentLoop) {
+  /// check that the loop has a single latch
+  if (!CurrentLoop.getLoopLatch())
+    return nullptr;
+
+  /// check that the loop has a single exit block
```

Comment:
(I believe here we want to find a single latch, a single exiting block, and they must be the same)

Reply:
in case of multiple exit blocks, `getExitingBlock` will return a null pointer, in which case I abort.

The loop latch will return the block that loops back to the header. 
I can verify that the Exiting Block dominates the Loop latch.

Reply:
I added a comment to clarify the exiting block check.